keps/sig-network/20200204-NetworkPolicy-verification-rearchitecture.md

```diff
@@ -0,0 +1,664 @@
+
+
+
+#### Logging verbosity is worse for slow tests
+ 
+CURRENTLY: Slow running tests are also hard to understand, because logging and metadata is expanded over a larger period of time, increasing the amount
+of information needed to be attended to diagnose an issue. For example, to test this, we have intentionally misconfigured my CIDR information for a calico CNI,
+and found that the following verbose logging about is returned when running the `NetworkPolicy` suite:
+```
+Feb  4 16:01:16.747: INFO: Pod "client-a-swm8q": Phase="Pending", Reason="", readiness=false. Elapsed: 1.87729ms
+... 26 more lines ...
+Feb  4 16:02:04.808: INFO: Pod "client-a-swm8q": Phase="Failed", Reason="", readiness=false. Elapsed: 48.063517483s
+```
+Thus, the majority of the logging information from a CNI which may have an issue is actually related to the various polling operations
+which occurred, rather then to the test itself.  Of course, this makes sense - since we always recreate pods, we have to potentially
+wait many seconds for those pods to come up.
+ 
+Thus, by increasing the performance of our tests, we also increase their understandability, because the amount of information needed to be
+audited for inspecting a failure may be reduced by 50% (currently, 50% of the output for failing network policy tests is that of the polling
+process for pods spinning up, which is easily avoided by a fixed server and client pod).
+ 
+### Documentation
+ 
+Documenting network states is very hard, in any scenario. Since the NetworkPolicy ginkgo tests are currently not documented outside of the code, no specific evidence is required here.  This proposal aims not to document these tests, but rather, to make the code more readable, and thus self-documenting.  However, formal documentation of how network policies, generally, are evaluated using a truth table approach, is part of this proposal.
+This generic documentation will be insightful and concise for those needing to test their NetworkPolicy implementations, and likely to not become obsolete, due to the generic nature of the truth-table/matrix approach (compared to the highly specific nature of existing tests).
+
+As a few examples of this:
+- the test `Creating a network policy for the server which allows traffic from the pod 'client-a' in same namespace` actually needs to confirm that *no outside* namespace can communicate with the server.
+- the test outlined in Figure2 is another example of a test which isn't described comprehensively.
+
+In the solutions section, we will highlight how the proposal makes these tests, and thus the semantics of network policies, explicit and self documenting.
+ 
+## Goals
+ 
+In short, our solution to this problem follows
```

Comment:
I would like to raise another point: It's hard to maintain network policies and be compliant: review changes to policies, audit them, etc. It will be nice to have an ability to use new API to test end user Infrastructure and include it as a library to infrastructure testing frameworks.

Reply:
Sure, good point - this is partially creation of a library so, agree, there is a lot of value esp. for CNI providers and people testing larger CNIs to reuse these bits.

- put the core library in a `kubernetes-sigs/` subrepository,
- or just have docs on how to import it using some vendoring or copying hacks 

any other thoughts  on wether sharing this as a library is feasible @bowei @caseydavenport ?

Reply:
in general i like this idea.. but maybe too much to chew on for this KEP? not sure if others share the same sentiment

Reply:
ok, lets get the tests into test/e2e, then we can work on `libraryfication` later.....